components/image-builder-mk3/pkg/auth/auth.go

```diff
@@ -108,9 +111,88 @@ func (a *DockerConfigFileAuth) Authenticate(registry string) (auth *Authenticati
 	}, nil
 }
 
+// CompositeAuth returns the first non-empty authentication of any of its consitutents
+type CompositeAuth []RegistryAuthenticator
+
+func (ca CompositeAuth) Authenticate(ctx context.Context, registry string) (auth *Authentication, err error) {
+	for _, ath := range ca {
+		res, err := ath.Authenticate(ctx, registry)
+		if err != nil {
+			return nil, err
+		}
+		if !res.Empty() {
+			return res, nil
+		}
+	}
+	return &Authentication{}, nil
+}
+
+func NewECRAuthenticator(ecrc *ecr.Client) *ECRAuthenticator {
+	return &ECRAuthenticator{
+		ecrc: ecrc,
+	}
+}
+
+type ECRAuthenticator struct {
+	ecrc *ecr.Client
+
+	ecrAuth                string
+	ecrAuthLastRefreshTime time.Time
+	ecrAuthLock            sync.Mutex
+}
+
+func (ath *ECRAuthenticator) Authenticate(ctx context.Context, registry string) (auth *Authentication, err error) {
+	// TODO(cw): find better way to detect if ref is an ECR repo
+	if !strings.Contains(registry, ".dkr.") || !strings.Contains(registry, ".amazonaws.com") {
+		return nil, nil
+	}
+
+	ath.ecrAuthLock.Lock()
+	defer ath.ecrAuthLock.Unlock()
+	// ECR tokens are valid for 12h: https://docs.aws.amazon.com/AmazonECR/latest/APIReference/API_GetAuthorizationToken.html
+	if time.Since(ath.ecrAuthLastRefreshTime) > 10*time.Hour {
```

Comment:
For the ecr-token-refresh cron job, we've made this 4 hours, to ensure we refresh at least twice before the full expiry. Do we want to do the same here? We've seen in the past that a refresh did fail

Reply:
Good call - will update